client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Module dependencies
+ */
+import config from 'config';
+import localforage from 'localforage';
+import Hashes from 'jshashes';
+import { isWhitelisted } from './whitelist-handler';
+
+import debugFactory from 'debug';
+const debug = debugFactory( 'calypso:sync-handler' );
+
+// defaults
+const defaults = config( 'sync-handler-defaults' );
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, wrappedResponseCallback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, wrappedResponseCallback );
+			}
+
+			const clonedParams = Object.assign( {}, params );
+			const { path } = clonedParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, wrappedResponseCallback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			this.retrieveRecord( key, ( err, localRecord ) => {
```

Comment:
This is a pretty long function. It may be easier to read promise style. 
```
this.retrieveRecord( key )
    .then( cacheCallback, recordErrorHandler )
    .then( networkFetch )
    .then( cacheResponse, networkErrorHandler )
    .then( networkCallback )
```